File: src/evaluation/Workspace.m
% The class Workspace is the outermost structure that stores optimizable variables.
% It represents the process of hashing in all levels (including the global one).
% For every level, it adds up num_block of all parts, and compute the number of retained
% triples. This is done by creating an optimizable variable instead of using min().

% Also, it manages all parts and the global stage.
% When Build() is called, it creates instances for parts and global stage recursively.
% When Evaluate() is called, it calls EvaluateInit(), EvaluatePre(), EvaluatePost() for the global stage & all parts.
% The returned value of Evaluate() is ALL constraints of the system. It consists two cell arrays [c, ceq] of GVar. The convertion will be done outside.

classdef Workspace < handle
properties
  num_retain_comp_id  % 1x3 cell array of 1 by max_level cell-array
  num_retain_comp  % 1x3 cell array of 1 by max_level cell-array of GVar
  num_retain_comp_lasteval
  num_retain_glob_id
  num_retain_glob  % scalar GVar
  num_retain_glob_lasteval
  single_mat_size_id
  single_mat_size  % should equal min(mat_size(1), mat_size(2), mat_size(3) / K)
  single_mat_size_lasteval
  value  % final value in Schonhage's inequality.
  omega_id
  omega
  K_id
  K

  % Auxiliary entries for performance
  num_constr_c
  num_constr_ceq
end
methods
  function obj = Workspace()
    % Do nothing.
  end

  function obj = Build(obj, max_level_)
    % Initialize the object and register the variables.
    global max_level param_manager parts globstage expinfo;
    max_level = max_level_;
    param_manager = ParamManager();  % Initialize the parameter manager.
    obj.num_retain_comp_id = cell(1, 3);
    obj.num_retain_glob_id = cell(1, 3);
    for r = 1 : 3
      obj.num_retain_comp_id{r} = cell(1, max_level);
      for lv = 2 : max_level
        obj.num_retain_comp_id{r}{lv} = param_manager.Register(1, 0, Inf, [0, 1 / 100]);
      end
      obj.num_retain_glob_id{r} = param_manager.Register(1, 0, Inf, [0, 1 / 100]);
    end
    obj.single_mat_size_id = param_manager.Register(1, 0, Inf, [0, 1 / 100]);
    
    % Register objectives.
    if strcmp(expinfo.obj_mode, 'alpha')
      % We restrict omega == 2. A negligible error of 1e-9 is allowed due to floating-number error.
      goal_eps = 1e-9;
      obj.omega_id = param_manager.Register(1, 2, 2 + goal_eps);
    else
      obj.omega_id = param_manager.Register(1, 0, Inf);
    end
    if strcmp(expinfo.obj_mode, 'omega')
      % We restrict K == expinfo.K.
      obj.K_id = param_manager.Register(1, expinfo.K, expinfo.K);
    else
      obj.K_id = param_manager.Register(1, 0, Inf);
    end
    if strcmp(expinfo.obj_mode, 'mu')
      % We restrict omega(K) <= 1 + 2 * K.
      param_manager.AddLinearConstraint({obj.omega_id, 1; obj.K_id, -2}, 1);
    end
    
    obj.num_constr_c = 0;
    obj.num_constr_ceq = 0;
    % Recursively build things.
    parts = cell(1, max_level);
    globstage = GlobalStage();
    globstage.Build();
  end

  function SetInitial(obj)
    % This function only sets the auxiliary variables in the black box.
    % Before calling, the user should set K and omega to desired values.
    global param_manager max_level;
    obj.Evaluate();  % Get evaluated auxiliary variables
    for r = 1 : 3
      for l = 2 : max_level
        param_manager.SetSingleParam(obj.num_retain_comp_id{r}{l}, obj.num_retain_comp_lasteval{r}{l});
      end
      param_manager.SetSingleParam(obj.num_retain_glob_id{r}, obj.num_retain_glob_lasteval{r});
    end
    param_manager.SetSingleParam(obj.single_mat_size_id, obj.single_mat_size_lasteval);
  end

  function [c, ceq] = Evaluate(obj)
    % Initialization Step: extract the values of the optimizable variables.
    global max_level param_manager parts globstage q;
    obj.num_retain_comp = cell(1, 3);
    obj.num_retain_glob = cell(1, 3);
    obj.num_retain_comp_lasteval = cell(1, 3);
    obj.num_retain_glob_lasteval = cell(1, 3);
    for r = 1 : 3
      obj.num_retain_comp{r} = cell(1, max_level);
      obj.num_retain_comp_lasteval{r} = cell(1, max_level);
      for l = 2 : max_level
        obj.num_retain_comp{r}{l} = param_manager.GetParam(obj.num_retain_comp_id{r}{l});
      end
      obj.num_retain_glob{r} = param_manager.GetParam(obj.num_retain_glob_id{r});
    end
    obj.single_mat_size = param_manager.GetParam(obj.single_mat_size_id);
    obj.omega = param_manager.GetParam(obj.omega_id);
    obj.K = param_manager.GetParam(obj.K_id);
    
    % Initialization for parts
    globstage.EvaluateInit();
    for l = 1 : max_level
      for i = 1 : length(parts{l})
        parts{l}{i}.EvaluateInit();
      end
    end

    % Propagate the fractions of parts
    globstage.EvaluatePre();
    for l = max_level : -1 : 1
      for i = 1 : length(parts{l})
        parts{l}{i}.EvaluatePre();
      end
    end

    % Evaluate the intermediate results
    for l = 1 : max_level
      for i = 1 : length(parts{l})
        parts{l}{i}.EvaluatePost();
      end
    end
    globstage.EvaluatePost();

    % Collect the nonlinear constraints.
    c = cell(1, obj.num_constr_c);
    ceq = cell(1, obj.num_constr_ceq);
    len_c = 0;
    len_ceq = 0;

    function AddC(c_now)
      for tmp_addc = 1 : length(c_now)
        len_c = len_c + 1;
        c{len_c} = c_now{tmp_addc};
      end
    end

    function AddCeq(ceq_now)
      for tmp_addceq = 1 : length(ceq_now)
        len_ceq = len_ceq + 1;
        ceq{len_ceq} = ceq_now{tmp_addceq};
      end
    end

    % Lagrange constraints
    for l = 1 : max_level
      for i = 1 : length(parts{l})
        % If is PartInfoLv2 or PartInfoZero, does not have Lagrange constraints.
        if isa(parts{l}{i}, 'PartInfo')
          ceq_now = parts{l}{i}.GetLagrangeConstraints();
          AddCeq(ceq_now);
        end
      end
    end
    ceq_now = globstage.GetLagrangeConstraints();
    AddCeq(ceq_now);

    % Asymmetric hashing: num_of_retain constraints
    for l = 3 : max_level  % Asymmetric hashing starts from level-3
      for r = 1 : 3
        % Component hashing from level-l to level-(l-1), region r
        num_block = GVar(param_manager.num_input, [0, 0, 0]);
        hash_penalty = GVar(param_manager.num_input, 0);
        p_comp = GVar(param_manager.num_input, 0);
        for i = 1 : length(parts{l})
          num_block = num_block + parts{l}{i}.num_block_contribution{r};  % The frac_part and region_prop within the part are already considered.
          hash_penalty = hash_penalty + parts{l}{i}.hash_penalty_term{r};
          p_comp = p_comp + parts{l}{i}.p_comp{r};
        end
        obj.num_retain_comp_lasteval{r}{l} = Inf;
        for t = 1 : 3
          if t ~= r
            AddC({obj.num_retain_comp{r}{l} - num_block(t) + hash_penalty});
            foo = num_block(t) - hash_penalty;
            obj.num_retain_comp_lasteval{r}{l} = min(obj.num_retain_comp_lasteval{r}{l}, foo.value);
          else
            AddC({obj.num_retain_comp{r}{l} - num_block(t) + p_comp});
            foo = num_block(t) - p_comp;
            obj.num_retain_comp_lasteval{r}{l} = min(obj.num_retain_comp_lasteval{r}{l}, foo.value);
          end
        end
      end
    end
    % Symmetric hashing at level 2
    % Note that num_block_contribution of level-2 is NOT a 1x3 cell array.
    % (It is a 1x3 GVar indicating the contribution to N_X, N_Y, and N_Z).
    num_block = GVar(param_manager.num_input, [0, 0, 0]);
    for i = 1 : length(parts{2})
      num_block = num_block + parts{2}{i}.num_block_contribution;
    end
    % obj.num_retain_comp{1}{2} (region 1, level 2) is utilized to represent the min() relationship.
    % {2}{2} and {3}{2} (i.e., regions 2 & 3) are omitted and always set to zero.
    for t = 1 : 3
      AddC({obj.num_retain_comp{1}{2} - num_block(t)});
    end
    AddCeq({obj.num_retain_comp{2}{2}});
    AddCeq({obj.num_retain_comp{3}{2}});
    obj.num_retain_comp_lasteval{1}{2} = min(num_block.value);
    obj.num_retain_comp_lasteval{2}{2} = 0;
    obj.num_retain_comp_lasteval{3}{2} = 0;
    
    % Global hashing from CW power to max_level
    for r = 1 : 3
      num_block = globstage.num_block{r};
      hash_penalty = globstage.hash_penalty_term{r};
      p_comp = globstage.p_comp{r};
      obj.num_retain_glob_lasteval{r} = Inf;
      for t = 1 : 3
        if t ~= r
          AddC({obj.num_retain_glob{r} - num_block(t) + hash_penalty});
          foo = num_block(t) - hash_penalty;
          obj.num_retain_glob_lasteval{r} = min(obj.num_retain_glob_lasteval{r}, foo.value);
        else
          AddC({obj.num_retain_glob{r} - num_block(t) + p_comp});
          foo = num_block(t) - p_comp;
          obj.num_retain_glob_lasteval{r} = min(obj.num_retain_glob_lasteval{r}, foo.value);
        end
      end
    end

    % Single Matrix Size constraints
    % See what is the largest <a, a, a^K> we can obtain.
    mat_size = globstage.mat_size;
    AddC({obj.single_mat_size - mat_size(1), ...
          obj.single_mat_size - mat_size(2), ...
          obj.single_mat_size - mat_size(3) ./ obj.K});
    obj.single_mat_size_lasteval = min([mat_size.value(1), mat_size.value(2), mat_size.value(3) ./ obj.K.value]);

    % value satisfies Schonhage
    % value equals the sum of num_retain plus mat_size * omega
    obj.value = obj.single_mat_size * obj.omega;
    for r = 1 : 3
      obj.value = obj.value + obj.num_retain_glob{r};  % Have considered region_frac.
    end
    for l = 2 : max_level
      for r = 1 : 3
        obj.value = obj.value + obj.num_retain_comp{r}{l};
      end
    end
    AddC({log(q + 2) * (2 .^ (max_level - 1)) - obj.value});  % value >= power * log(q + 2)

    % Set back the auxiliary variables
    % len_c and len_ceq are unnecessary auxiliary variables but could avoid dynamic allocation.
    if length(c) > len_c
      c = c(1 : len_c);
    end
    if length(ceq) > len_ceq
      ceq = ceq(1 : len_ceq);
    end
    obj.num_constr_c = len_c;
    obj.num_constr_ceq = len_ceq;
  end

  function omega_pos = GetOmegaPos(obj)
    % [Hacky function] Return the position of omega in the param_manager.
    global param_manager;
    omega_pos = param_manager.group_startpos(obj.omega_id);
  end

  function K_pos = GetKPos(obj)
    % [Hacky function] Return the position of K in the param_manager.
    global param_manager;
    K_pos = param_manager.group_startpos(obj.K_id);
  end
end
end

File: fast_mm_python/src/evaluation/Workspace.py
"""
Workspace class - the outermost structure for managing optimization variables.

Represents the process of hashing at all levels and manages all parts and the global stage.
"""

import numpy as np
from ..autograd import ParamManager, GVar


# Global workspace instance
_workspace = None


def get_workspace():
    """Get the global workspace instance."""
    global _workspace
    return _workspace


def set_workspace(ws):
    """Set the global workspace instance."""
    global _workspace
    _workspace = ws


class Workspace:
    """
    Workspace manages all optimization variables and constraint evaluation.
    
    It creates instances for parts and global stage recursively and
    evaluates all constraints of the system.
    """
    
    def __init__(self):
        """Initialize empty workspace."""
        self.num_retain_comp_id = None
        self.num_retain_comp = None
        self.num_retain_comp_lasteval = None
        self.num_retain_glob_id = None
        self.num_retain_glob = None
        self.num_retain_glob_lasteval = None
        self.single_mat_size_id = None
        self.single_mat_size = None
        self.single_mat_size_lasteval = None
        self.value = None
        self.omega_id = None
        self.omega = None
        self.K_id = None
        self.K = None
        
        self.num_constr_c = 0
        self.num_constr_ceq = 0
        
        self.max_level = None
        self.param_manager = None
        self.parts = None  # Dictionary: {level: {id: part}}
        self.globstage = None
    
    def Build(self, max_level_):
        """
        Initialize the workspace and register variables.
        Build the complete PartInfo hierarchy recursively.
        
        Args:
            max_level_: Maximum level for the computation
        """
        from ..autograd import set_param_manager
        from ..control import get_expinfo
        from .GlobalStage import GlobalStage
        
        self.max_level = max_level_
        max_level = max_level_
        
        # Initialize parameter manager
        self.param_manager = ParamManager()
        set_param_manager(self.param_manager)
        set_workspace(self)
        
        expinfo = get_expinfo()
        
        # Register num_retain_comp variables
        self.num_retain_comp_id = [None, None, None]
        self.num_retain_glob_id = [None, None, None]
        
        for r in range(3):
            self.num_retain_comp_id[r] = [None] * (max_level + 1)
            for lv in range(2, max_level + 1):
                self.num_retain_comp_id[r][lv] = self.param_manager.Register(
                    1, lb=0, ub=np.inf, initializer=(0, 0.01)
                )
            self.num_retain_glob_id[r] = self.param_manager.Register(
                1, lb=0, ub=np.inf, initializer=(0, 0.01)
            )
        
        self.single_mat_size_id = self.param_manager.Register(
            1, lb=0, ub=np.inf, initializer=(0, 0.01)
        )
        
        # Register objectives based on mode
        obj_mode = expinfo.get('obj_mode', 'omega')
        
        if obj_mode == 'alpha':
            # Restrict omega == 2
            goal_eps = 1e-9
            self.omega_id = self.param_manager.Register(1, lb=2.0, ub=2.0 + goal_eps)
        else:
            self.omega_id = self.param_manager.Register(1, lb=0, ub=np.inf)
        
        if obj_mode == 'omega':
            # Restrict K == expinfo.K
            K_val = expinfo.get('K', 1.0)
            self.K_id = self.param_manager.Register(1, lb=K_val, ub=K_val)
        else:
            self.K_id = self.param_manager.Register(1, lb=0, ub=np.inf)
        
        if obj_mode == 'mu':
            # Restrict omega(K) <= 1 + 2 * K
            self.param_manager.AddLinearConstraint([
                (self.omega_id, np.array([[1.0]])),
                (self.K_id, np.array([[-2.0]]))
            ], np.array([1.0]))
        
        # Initialize parts dictionary
        self.parts = {}
        
        # Initialize global stage and build recursively
        self.globstage = GlobalStage()
        self.globstage.Build(max_level, self.parts)
        
        # print(f"[INFO] Workspace built with {self.param_manager.num_input} parameters")
        # print(f"[INFO] Parts created at levels 2-{max_level}")
        for level in range(2, max_level + 1):
            if level in self.parts:
                print(f"       Level {level}: {len(self.parts[level])} parts")
    
    def SetInitial(self):
        """
        Set initial values for auxiliary variables.
        Before calling, user should set K and omega to desired values.
        """
        # Evaluate to get auxiliary variable values
        self.Evaluate()
        
        # Set auxiliary variables
        for r in range(3):
            for l in range(2, self.max_level + 1):
                self.param_manager.SetSingleParam(
                    self.num_retain_comp_id[r][l],
                    self.num_retain_comp_lasteval[r][l]
                )
            self.param_manager.SetSingleParam(
                self.num_retain_glob_id[r],
                self.num_retain_glob_lasteval[r]
            )
        
        self.param_manager.SetSingleParam(
            self.single_mat_size_id,
            self.single_mat_size_lasteval
        )
    
    def Evaluate(self):
        """
        Evaluate all constraints.
        
        This is the core evaluation function that:
        1. Loads all parameters into GVars
        2. Evaluates all parts in three phases (Init, Pre, Post)
        3. Evaluates the global stage
        4. Collects all constraints
        
        Returns:
            c, ceq: Lists of GVar instances representing inequality and equality constraints
        """
        from ..control import get_q
        
        q = get_q()
        
        # Extract optimizable variables
        self.num_retain_comp = [None, None, None]
        self.num_retain_glob = [None, None, None]
        self.num_retain_comp_lasteval = [None, None, None]
        self.num_retain_glob_lasteval = [None, None, None]
        
        for r in range(3):
            self.num_retain_comp[r] = [None] * (self.max_level + 1)
            self.num_retain_comp_lasteval[r] = [None] * (self.max_level + 1)
            
            for lv in range(2, self.max_level + 1):
                var = self.param_manager.GetParam(self.num_retain_comp_id[r][lv])
                self.num_retain_comp[r][lv] = var
                self.num_retain_comp_lasteval[r][lv] = var.value[0]
            
            var = self.param_manager.GetParam(self.num_retain_glob_id[r])
            self.num_retain_glob[r] = var
            self.num_retain_glob_lasteval[r] = var.value[0]
        
        var = self.param_manager.GetParam(self.single_mat_size_id)
        self.single_mat_size = var
        self.single_mat_size_lasteval = var.value[0]
        
        self.omega = self.param_manager.GetParam(self.omega_id)
        self.K = self.param_manager.GetParam(self.K_id)
        
        # Initialize constraint lists
        c = []
        ceq = []
        
        # Phase 1: EvaluateInit - Initialize all parts
        for level in range(2, self.max_level + 1):
            if level in self.parts:
                for part in self.parts[level].values():
                    part.EvaluateInit()
        
        self.globstage.EvaluateInit()
        
        # Phase 2: EvaluatePre - Propagate part_frac from top to bottom
        self.globstage.EvaluatePre()
        
        for level in range(self.max_level, 1, -1):  # From max_level down to 2
            if level in self.parts:
                for part in self.parts[level].values():
                    if hasattr(part, 'EvaluatePre'):
                        part.EvaluatePre()
        
        # Phase 3: EvaluatePost - Compute contributions from bottom to top
        for level in range(2, self.max_level + 1):  # From 2 up to max_level
            if level in self.parts:
                for part in self.parts[level].values():
                    if hasattr(part, 'EvaluatePost'):
                        part.EvaluatePost()
        
        self.globstage.EvaluatePost(self.parts)
        
        # Collect constraints from parts
        
        # Lagrange constraints (if needed for optimization)
        # for level in range(3, self.max_level + 1):
        #     if level in self.parts:
        #         for part in self.parts[level].values():
        #             if hasattr(part, 'GetLagrangeConstraints'):
        #                 ceq.extend(part.GetLagrangeConstraints())
        
        # Global stage Lagrange constraints
        # if hasattr(self.globstage, 'GetLagrangeConstraints'):
        #     ceq.extend(self.globstage.GetLagrangeConstraints())
        
        return c, ceq
    
    def GetOmegaPos(self):
        """Get the position of omega in the parameter vector."""
        return self.param_manager.group_startpos[self.omega_id]
    
    def GetKPos(self):
        """Get the position of K in the parameter vector."""
        return self.param_manager.group_startpos[self.K_id]
    
    def GetObjective(self):
        """
        Compute the objective value based on the mode.
        
        Returns:
            objective: Scalar value to minimize
        """
        from ..control import get_expinfo
        expinfo = get_expinfo()
        obj_mode = expinfo.get('obj_mode', 'omega')
        
        if obj_mode == 'omega':
            # Minimize omega
            return self.omega.value[0]
        elif obj_mode == 'alpha':
            # Maximize K (minimize -K)
            return -self.K.value[0]
        elif obj_mode == 'mu':
            # Minimize K
            return self.K.value[0]
        else:
            raise ValueError(f"Unknown objective mode: {obj_mode}")
